File: 02_kafka_consumer_examples/scalas/KafkaTwitterConsumerExample/src/main/scala/com/dimanche/examples/KafkaTwitterConsumerExample.scala
package com.dimanche.examples

import spray.json._
import DefaultJsonProtocol._

import java.util.Properties
import java.util.Collections.singletonList

import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.kafka.clients.consumer.{ConsumerRecords, KafkaConsumer}

object KafkaTwitterConsumerExample {

    val TOPIC : String  = "topic_one"

    val props : Properties = new Properties()
    props.put("bootstrap.servers", "localhost:9092")
    props.put ("auto.offset.reset", "earliest" )
    props.put ("group.id", "topic_one" )
    props.put("key.deserializer", new StringDeserializer().getClass.getName)
    props.put("value.deserializer", new StringDeserializer().getClass.getName)


    def main(args: Array[String]): Unit = {

        val kafkaConsumer = new KafkaConsumer[String, String](props)
        kafkaConsumer.subscribe(singletonList(TOPIC))

        while(true){

            val records : ConsumerRecords[String,String] = kafkaConsumer.poll(2000)// poll is currently deprecated
            val recordsIter = records.iterator()

            while (recordsIter.hasNext){
                val record = recordsIter.next
                val currentOffset = record.offset()
                println(currentOffset)
                try {
                    val jsRecord = record.value().parseJson.asJsObject()
                    val data = (jsRecord.getFields("user"),jsRecord.getFields("lang"),
                      jsRecord.getFields("id"),jsRecord.getFields("text"),
                      jsRecord.getFields("created_at"))
                    println(currentOffset, data)
                } finally {
                    println("finally")
                }
            }
        }
    }
}

File: 02_kafka_consumer_examples/java/KafkaTwitterConsumerExample/src/main/java/com/dimanche/examples/KafkaTwitterConsumerExample.java
package com.dimanche.examples;

import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.common.serialization.LongDeserializer;
import org.apache.kafka.common.serialization.StringDeserializer;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Properties;


public class KafkaTwitterConsumerExample {

    private static KafkaConsumer<String, String> createConsumer() {
        final Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,     Constants.BOOTSTRAP_SERVER);
        props.put(ConsumerConfig.GROUP_ID_CONFIG,  Constants.GROUP_ID);
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,   StringDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,  StringDeserializer.class.getName());

        return new KafkaConsumer<>(props);
    }

    public static void main(String[] args) throws InterruptedException, IOException {

        final KafkaConsumer<String, String> kafkaConsumer = createConsumer();
        kafkaConsumer.subscribe(Collections.singleton(Constants.TOPIC));

        Constants.TOPICS.add("topic_one");

        kafkaConsumer.subscribe(Constants.TOPICS);

        String data=null;

        while (true){
            try{
                ConsumerRecords<String, String> records = kafkaConsumer.poll(10);
                for (ConsumerRecord<String, String> record: records){

                    System.out.println(String.format("Topic - %s, Partition - %d, \n Value: %s", record.topic(), record.partition(), data));
                }
            }catch (Exception e){
                System.out.println(e.getMessage());
            }
        }

//        kafkaConsumer.close();
//        }
    }
}
